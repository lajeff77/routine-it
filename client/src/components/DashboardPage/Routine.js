import React, { useEffect, useState } from 'react';
import axios from "axios";
import {Box, Button, LinearProgress, ListItem, Typography} from '@mui/material';
import {useNavigate} from 'react-router-dom';

const Routine = (props) =>{
    const [mode,setMode] = useState("");
    const [tasks, setTasks] = useState([]);
    const [progress, setProgress] = useState(0);

    const navigate = useNavigate();

    useEffect(() =>{
        getTasks();
        //set mode initially when component loads
        setMode(props.mode);

    },[]);

    useEffect(()=>{
        //calc progress
        setProgress(calculateProgress);
    },[tasks])


    useEffect(() => {
            const putData = async() =>{
                try{
                    props.routine.mode = mode;
                    await axios.put(`http://localhost:8080/api/v1/routines/${props.id}`, props.routine);
                }catch(error){
                    console.log(`An error occurred while updating the routine: ${error}`);
            }
            
        }
        putData();

        //re calc progress
        setProgress(calculateProgress);
    },[mode]);

    const getTasks = async () => {
        try{
            const response = await axios.get(`http://localhost:8080/api/v1/tasks/${props.id}`);
            setTasks(response.data)
        
        }catch(error){
            console.log(`an error occurred while fetching the tasks to calculate the percentage: ${error}`)
        }
    }

    const calculateProgress = () =>{
        //sort tasks
        let totalTasks = 0;
        let completedTasks = 0;
        for(let i = 0; i < tasks.length; i++)
        {
            if(tasks[i].highEffort && mode === "HIGH")
            {
                totalTasks++;
                if(tasks[i].complete)
                    completedTasks++;
            }
                
            if(tasks[i].mediumEffort && mode === "MEDIUM")
            {
                totalTasks++;
                if(tasks[i].complete)
                    completedTasks++;
            }
                
            if(tasks[i].lowEffort && mode === "LOW")
            {
                totalTasks++;
                if(tasks[i].complete)
                    completedTasks++;
            }
                
        }
        let percent = (completedTasks/totalTasks) *100;
        return isNaN(percent) ? 0 : percent;
    }

    const determineLabel = () => {
        if(progress === 0)
            return "Start"
        else {
            if(progress === 100)
                return "Completed"
            else return "Continue"
        }
    }

    const navigateToRoutineChecklist = () => {
        navigate("/routinechecklist",{state: {id: props.id}});
    }

    return(
        <ListItem sx={{margin:"10px", display:"flex", justifyContent:"space-around"}}>
            <Typography sx={{width:"150px"}}>{props.name}</Typography>
            <Box sx={{width:"40%"}}><LinearProgress variant="determinate" value={progress} sx={{borderRadius:"5px", margin:"5px"}}/></Box>
            <Box sx={{display:"flex", flexDirection: 'column', justifyContent:'space-around'}}>
                <Typography sx={{textAlign: 'center'}}>Effort Level</Typography>
                <Box sx={{display:"flex", justifyContent:'space-around'}}>
                    <Button type="button" color="error"  variant={mode === "HIGH" ? "contained": "outlined"} sx={{margin:"5px", width:"75px"}} onClick={() => {setMode("HIGH")}}>HIGH</Button>
                    <Button type="button" color="warning" variant={mode === "MEDIUM" ? "contained": "outlined"} sx={{margin:"5px", width:"75px"}} onClick={() => {setMode("MEDIUM")}}>MEDIUM</Button>
                    <Button type="button" color="success" variant={mode === "LOW" ? "contained": "outlined"} sx={{margin:"5px", width:"75px"}} onClick={() => {setMode("LOW")}}>LOW</Button>
                </Box>
            </Box>
          
            <Button type="button" variant="outlined" sx={{margin:"5px", width:"200px"}} onClick={() => {navigateToRoutineChecklist()}}>{determineLabel()}</Button>
        </ListItem>
    );
};

export default Routine;