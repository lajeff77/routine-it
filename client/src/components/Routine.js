import React from 'react';
import {Box, Button, LinearProgress, ListItem, Typography} from '@mui/material';
import {useNavigate} from 'react-router-dom';

const Routine = (props) =>{
    let label = determineLabel(props.progress);
    const navigate = useNavigate();

    const navigateToRoutineChecklist = () => {
        console.log('hi');
        navigate("/routinechecklist");
    }

    return(
        <ListItem sx={{margin:"10px", display:"flex", justifyContent:"space-around"}}>
            <Typography sx={{width:"200px"}}>{props.name}</Typography>
            <Box sx={{width:"50%"}}><LinearProgress variant="determinate" value={Number(props.progress)} sx={{borderRadius:"5px"}}/></Box>
            <Button type="button" variant="outlined" sx={{margin:"5px", width:"200px"}} onClick={navigateToRoutineChecklist}>{label}</Button>
        </ListItem>
    );
};

function determineLabel(progress){
    if(progress === "0")
        return "Start"
    else {
        if(progress === "100")
            return "Completed"
        else return "Continue"
    }
}

export default Routine;