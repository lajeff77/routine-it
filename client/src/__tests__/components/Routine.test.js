import { render } from "@testing-library/react";
import {expect, test} from '@jest/globals';
import React from "react";
import Routine from "../../components/Routine";


describe(Routine.name, ()=>{
    test('Routine matches snapshot', () => {
        const {container} = render(<Routine/>);
        expect(container.firstChild).toMatchSnapshot();
    });
});