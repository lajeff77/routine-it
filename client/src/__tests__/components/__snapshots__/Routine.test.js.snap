// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routine Routine matches snapshot 1`] = `
<li
  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-e4rxub-MuiListItem-root"
>
  <p
    class="MuiTypography-root MuiTypography-body1 css-vybajm-MuiTypography-root"
  />
  <div
    class="MuiBox-root css-11ze7cv"
  >
    <span
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="NaN"
      class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-ku8bnv-MuiLinearProgress-root"
      role="progressbar"
    >
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
        style="transform: translateX(NaN%);"
      />
    </span>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-128u0p7-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Continue
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</li>
`;
